hudi-common/src/main/java/org/apache/hudi/metadata/BaseFileRecordParsingUtils.java

```diff
@@ -0,0 +1,186 @@
+   * @throws IOException
+   */
+  @VisibleForTesting
+  public static Map<RecordStatus, List<String>> getRecordKeyStatuses(String basePath,
+                                                              String partition,
+                                                              String latestFileName,
+                                                              String prevFileName,
+                                                              HoodieStorage storage,
+                                                              Set<RecordStatus> recordStatusesOfInterest) throws IOException {
+    Set<String> recordKeysFromLatestBaseFile = getRecordKeysFromBaseFile(storage, basePath, partition, latestFileName);
+    if (prevFileName == null) {
+      if (recordStatusesOfInterest.contains(RecordStatus.INSERT)) {
+        return Collections.singletonMap(RecordStatus.INSERT, new ArrayList<>(recordKeysFromLatestBaseFile));
+      } else {
+        // if this is a new base file for a new file group, everything is an insert.
+        return Collections.emptyMap();
+      }
+    } else {
+      // read from previous base file and find difference to also generate delete records.
+      // we will return updates and deletes from this code block
+      Set<String> recordKeysFromPreviousBaseFile = getRecordKeysFromBaseFile(storage, basePath, partition, prevFileName);
+      Map<RecordStatus, List<String>> toReturn = new HashMap<>(recordStatusesOfInterest.size());
+      if (recordStatusesOfInterest.contains(RecordStatus.DELETE)) {
+        toReturn.put(RecordStatus.DELETE, recordKeysFromPreviousBaseFile.stream()
+            .filter(recordKey -> {
+              // deleted record
+              return !recordKeysFromLatestBaseFile.contains(recordKey);
+            }).collect(toList()));
+      }
+
+      List<String> updates = new ArrayList<>();
+      List<String> inserts = new ArrayList<>();
+      // do one pass and collect list of inserts and updates.
+      recordKeysFromLatestBaseFile.stream().forEach(recordKey -> {
+        if (recordKeysFromPreviousBaseFile.contains(recordKey)) {
+          updates.add(recordKey);
+        } else {
+          inserts.add(recordKey);
+        }
+      });
+      if (recordStatusesOfInterest.contains(RecordStatus.UPDATE)) {
+        toReturn.put(RecordStatus.UPDATE, updates);
+      }
+      if (recordStatusesOfInterest.contains(RecordStatus.INSERT)) {
+        toReturn.put(RecordStatus.INSERT, inserts);
+      }
+      return toReturn;
+    }
+  }
+
+  private static Set<String> getRecordKeysFromBaseFile(HoodieStorage storage, String basePath, String partition, String fileName) throws IOException {
+    StoragePath dataFilePath = new StoragePath(basePath, StringUtils.isNullOrEmpty(partition) ? fileName : (partition + Path.SEPARATOR) + fileName);
+    FileFormatUtils fileFormatUtils = HoodieIOFactory.getIOFactory(storage).getFileFormatUtils(HoodieFileFormat.PARQUET);
```

Comment:
I'm wondering is it safe to hardcode as `HoodieFileFormat.PARQUET`? Shouldn't the file format be inferred dynamically based on `basePath`?

Reply:
you are right, it should already be addressed?

Reply:
https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/metadata/BaseFileRecordParsingUtils.java#L172

Reply:
cc @nsivabalan for visibility.

Reply:
yes, we can put out a patch. I can get that out by tonight